### PR TITLE
Fixed leak in menu.js file. Fixes #12765

### LINKED ISSUE #12765: leaks in src/command/Menus.js
`Menu.prototype.addMenuItem(CMD_ID)` (1) calls constructor MenuItem(id, command) (2) which attaches event handlers to command (3). Calling `Menu.prototype.removeMenuItem(CMD_ID)` (4) doesn't remove these handlers leading to infinitely increasing numbers of event handlers on command.

1) https://github.com/adobe/brackets/blob/master/src/command/Menus.js#L546
2) https://github.com/adobe/brackets/blob/master/src/command/Menus.js#L586
3) https://github.com/adobe/brackets/blob/master/src/command/Menus.js#L295-L316
4) https://github.com/adobe/brackets/blob/master/src/command/Menus.js#L432-L466

One example of this would be https://github.com/zaggino/brackets-git/blob/master/src/Main.ts#L173-L203

### PR DESCRIPTION
**Pull request by daytonjallen**
Fixes https://github.com/adobe/brackets/issues/12765

### CODE REVIEW

**Comment by ficristo** [comment]
I know it is a bit confusing, but please, revert the changes to `src/config.json`.

**Comment by daytonallen** [comment]
I don't remember touching that file but I've reverted the changes. Sorry about that. Now the test is failing, looking into that now.

**Comment by daytonallen** [comment]
Checking the CI logs, I think the reason for the build failure isn't anything on my end. Right?

**Comment by MarcelGerber** [comment]
Yes, the build failure was on our side. It passed now :thumbsup:

cc @zaggino for review

**Comment by zaggino** [comment]
@daytonallen the code is great, but is missing a test and that should be done in this case as test for this is easy to write, please add it somewhere here: https://github.com/adobe/brackets/blob/master/test/spec/Menu-test.js#L660-L737

You should write it in a way that it fails when you comment out the `removeMenuItemEventListeners(commandObj);` call you added and passes when you uncomment it.

You'll need to compare contents of `._eventHandlers[eventName].length` before and after calling `removeMenuItem` in the test.

**Comment by daytonallen** [comment]
How do I run the tests locally to check if it's done correctly? I tried the `debug`'s menu `run tests` option but it doesn't seem to be detecting a new test case.

**Comment by zaggino** [comment]
@daytonallen did you setup your installed Brackets with your cloned Brackets folder like in here https://github.com/adobe/brackets/wiki/How-to-Hack-on-Brackets ? See the `setup_for_hacking` thingy.

**Comment by zaggino** [comment]
If you did, please push the testcase here, maybe something else is wrong with it, you can still correct it with additional commits.

**Comment by daytonallen** [comment]
Yeah I did

**Comment by zaggino** [comment]
@daytonallen you need to look for it here, it runs for me:
image
image

**Comment by daytonallen** [comment]
Turns out I was testing under the wrong grouping.

File: test/spec/Menu-test.js [hunk 1]
```diff
@@ -734,6 +734,39 @@ define(function (require, exports, module) {
...
+                    expect(menuItem).toBeDefined();
+
+                    $listItems = menuDOMChildren(menuItemId);
+                    expect($listItems.length).toBe(1);
+
+                    var command = CommandManager.get(commandId);
+                    expect(typeof (command)).toBe("object");
+
+                    // Test event
+                    command.on("nameChange", menuItem._nameChanged);
```
Comment (zaggino) [inline test/spec/Menu-test.js]: you shouldn't add event manually with `.on`, you should just do `expect` for events that are added by `addMenuItem`

**Comment by zaggino** [comment]
@daytonallen 
You can remove stuff copied from other tests, it should be somthing simple like
```
menu.addMenuItem(commandId);
expect(command._eventHandlers....);
menu.addMenuItem(commandId);
expect(command._eventHandlers....);
```
everything else doesn't need to be there

File: src/command/Menus.js [hunk 2]
```diff
@@ -179,6 +179,19 @@ define(function (require, exports, module) {
     }
 
     /**
+    * Removes the attached event listeners from the corresponding object.
+    * @param {Command} command
+    */
+    function removeMenuItemEventListeners(command) {
+        command
+            .off("enabledStateChange", command._enabledChanged)
```
Comment (zaggino) [inline src/command/Menus.js]: looking at the code now, `_enabledChanged` and other methods live on `MenuItem` not on `command` ... are you sure that you're not calling `.off("enabledStateChange", undefined)` ???

File: test/spec/Menu-test.js [hunk 3]
```diff
@@ -734,6 +734,35 @@ define(function (require, exports, module) {
...
+                runs(function () {
+                    var menuItemId = "menu-test-removeMenuItem4";
+                    var commandId = "Menu-test.removeMenuItem.command4";
+                    CommandManager.register("Brackets Test Command Custom", commandId, function () {});
+                    var menu = Menus.addMenu("Custom", menuItemId);
+
+
+                    var menuItem = menu.addMenuItem(commandId);
+
+                    var command = CommandManager.get(commandId);
```
Comment (zaggino) [inline test/spec/Menu-test.js]: to be sure you're detaching only stuff that's been added by `addMenuItem` I'd suggest adding a custom event here like this: `command.on("nameChange", function () {})` and that handler should stay attached even after you call `removeMenuItem`

**Comment by zaggino** [comment]
Thanks @daytonallen merged to master! :+1:

**Comment by daytonallen** [comment]
My pleasure. Thanks for all the guidance also.

**Pull request merged**